### PR TITLE
Pull request for One Note Url changes in the search pages and Exernal Sharing API changes

### PR DESCRIPTION
**Pull request by premchandpl**
Pull request for One Note Url changes in the search pages
Exernal Sharing API changes

### CODE REVIEW

File: tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Models/Matter/MatterInformationVM.cs [hunk 1]
```diff
@@ -7,6 +7,7 @@ public class MatterInformationVM: MatterVM
         public bool IsConflictCheck { get; set; }
         public string MatterLocation { get; set; }
         public string MatterCreator { get; set; }
-        public string MatterCreatorEmail { get; set; }                
+        public string MatterCreatorEmail { get; set; }        
+        public string RequestedUrl { get; set; }
```
Comment (premchandpl) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Models/Matter/MatterInformationVM.cs]: Added new property to check whether one note url exists or not

File: tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Models/Search/EntityData.cs [hunk 2]
```diff
@@ -25,5 +25,9 @@ public class EntityData
         /// Gets or sets the email of user or security group
         /// </summary>
         public string PrincipalType { get; set; }
+        public string SPUserID { get; set; }
```
Comment (premchandpl) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Models/Search/EntityData.cs]: Added new properties to support new external api

File: tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/Matter/IMatterRepository.cs [hunk 3]
```diff
@@ -76,5 +76,7 @@ public interface IMatterRepository: ICommonRepository
         /// <param name="matterConfigurations"></param>
         /// <returns></returns>
         GenericResponseVM SaveConfigurations(ClientContext clientContext, MatterConfigurations matterConfigurations);
+
+        bool OneNoteUrlExists(MatterInformationVM matterInformation);
```
Comment (premchandpl) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/Matter/IMatterRepository.cs]: Added new method to check one note url existence

File: tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/Matter/MatterRepository.cs [hunk 4]
```diff
@@ -228,8 +228,7 @@ public GenericResponseVM UpdateMatter(MatterInformationVM matterInformation)
                     MatterTask = matter.Name + matterSettings.TaskNameSuffix,
                     MatterSitePages = matterSettings.MatterLandingPageRepositoryName
                 };
-                RevertMatterUpdates(client, matter, clientContext, matterRevertListObject, loggedInUserName,
-                    userPermissionOnLibrary, listItemId, isEditMode);
+                RevertMatterUpdates(client, matter, matterRevertListObject, isEditMode, userPermissionOnLibrary);
```
Comment (premchandpl) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/Matter/MatterRepository.cs]: Reduced the number of input parameters

File: tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/Matter/MatterRepository.cs [hunk 5]
```diff
@@ -1312,13 +1311,20 @@ public void SetPropertBagValuesForList(ClientContext clientContext, PropertyValu
         /// <param name="matterLandingPageId">List item id</param>
         /// <param name="isEditMode">Add/ Edit mode</param>
         /// <returns>Status of operation</returns>
-        public bool RevertMatterUpdates(Client client, Matter matter, ClientContext clientContext, 
-            MatterRevertList matterRevertListObject, string loggedInUserTitle, IEnumerable<RoleAssignment> oldUserPermissions, 
-            int matterLandingPageId, bool isEditMode)
+        public bool RevertMatterUpdates(Client client, Matter matter, MatterRevertList matterRevertListObject, bool isEditMode, IEnumerable<RoleAssignment> oldUserPermissionOnLibrary
```
Comment (premchandpl) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/Matter/MatterRepository.cs]: Reduced the number of input parameters

File: tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/SharePointInfo/ExternalSharing.cs [hunk 6]
```diff
@@ -110,35 +109,60 @@ private GenericResponseVM SendExternalNotification(MatterInformationVM matterInf
...
-                        userRole.Role = SharePoint.Client.Sharing.Role.View;
+                        roleId = "1073741826";
                         break;
-                }
-                users.Add(userRole);                
-                #region Doc Sharing API
-                //Need to use MatterGuid instead of Id
+                }               
+
+                PeoplePickerUser peoplePickerUser = new PeoplePickerUser();
```
Comment (premchandpl) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/SharePointInfo/ExternalSharing.cs]: To support new external sharing API need to construct this new object

File: tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/SharePointInfo/ExternalSharing.cs [hunk 7]
```diff
@@ -110,35 +109,60 @@ private GenericResponseVM SendExternalNotification(MatterInformationVM matterInf
...
+                string emailSubject = $"The following information has been shared with you {matterInformation.Matter.Name}";
+                #region Doc Sharing API                
+                SharingResult sharingResult = null;                
                 using (var clientContext = spoAuthorization.GetClientContext(clientUrl))
                 {
-                    //Send notification to the matter landing page url with permission the user has selected
-                    IList<UserSharingResult> matterLandingPageResult = DocumentSharingManager.UpdateDocumentSharingInfo(clientContext,
-                    matterLandingPageUrl,
-                    users, true, true, true, "The following matter page has been shared with you", true, true);
+                    sharingResult = Web.ShareObject(clientContext, url, peoplePicker, roleValue, 0, propagateAcl, sendEmail, includeAnonymousLinkInEmail, emailSubject, emailBody, useSimplifiedRoles);
```
Comment (premchandpl) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/SharePointInfo/ExternalSharing.cs]: Used new API for sharing matter information to external user

File: tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/SharePointInfo/Interfaces/ISearch.cs [hunk 8]
```diff
@@ -18,5 +18,6 @@ public interface ISearch
         List<RoleDefinition> GetWebRoleDefinitions(Client client);
         IList<PeoplePickerUser> SearchUsers(SearchRequestVM searchRequestVM);
         GenericResponseVM GetConfigurations(string siteCollectionUrl, string listName);
+        bool PageExists(Client client, string requestedUrl);
```
Comment (premchandpl) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/SharePointInfo/Interfaces/ISearch.cs]: to check one note url existence

File: tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/SharePointInfo/Search.cs [hunk 9]
```diff
@@ -1271,10 +1271,58 @@ public static object GetPropertyValue(object obj, string name)
...
 
 
 
+        /// <summary>
+        /// Checks if the requested page exists or not.
+        /// </summary>
+        /// <param name="requestedUrl">URL of the page, for which check is to be performed</param>
+        /// <param name="clientContext">ClientContext for SharePoint</param>
+        /// <returns>true or false string based upon the existence of the page, referred in requestedUrl</returns>
+        public bool PageExists(Client client, string requestedUrl)
```
Comment (premchandpl) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/SharePointInfo/Search.cs]: to check for one note url existence

File: tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Web/Common/MatterProvision.cs [hunk 10]
```diff
@@ -527,16 +528,7 @@ public GenericResponseVM UpdateMatter(MatterInformationVM matterInformation)
             }
             catch (Exception ex)
             {
-                MatterRevertList matterRevertListObject = new MatterRevertList()
```
Comment (premchandpl) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Web/Common/MatterProvision.cs]: not necessary

File: tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Web/Controllers/MatterController.cs [hunk 11]
```diff
@@ -1213,6 +1206,47 @@ public IActionResult ShareMatter([FromBody] MatterInformationVM matterInformatio
...
+        /// <param name="matterInformation"></param>
+        /// <returns></returns>
+        [HttpPost("onenoteurlexists")]
+        [Produces(typeof(GenericResponseVM))]
+        [SwaggerOperation("onenoteurlexists")]
+        [SwaggerResponse(HttpStatusCode.OK,
+            Description = "This method will check whether one note url exists for a selected matter or not",
+            Type = typeof(GenericResponseVM))]
+        [SwaggerResponseRemoveDefaults]
+        public IActionResult OneNoteUrlExists([FromBody]MatterInformationVM matterInformation)
```
Comment (premchandpl) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Web/Controllers/MatterController.cs]: Added new web api to check whether the matter is having one note document

File: tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Web/wwwroot/app/dashboard/settings.controller.js [hunk 12]
```diff
@@ -118,8 +120,10 @@
             vm.showClientDetails = false;
             //#endregion
 
+
+            
             var optionsForClientUrl = {                
-                    Url: configs.global.repositoryUrl                
+                Url: configs.global.isBackwardCompatible ? configs.global.repositoryUrl : configs.uri.SPOsiteURL
```
Comment (premchandpl) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Web/wwwroot/app/dashboard/settings.controller.js]: to fix the settings issue

File: tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Web/wwwroot/app/matter/MatterTemplates/MatterCellTemplate.html [hunk 13]
```diff
@@ -10,18 +10,22 @@
...
-
-                    <a title={{grid.appScope.vm.matterConfigContent.ECBMenu3Title}} class='ECBItem ms-ContextualMenu-link upload' href='{{row.entity.matterClientUrl}}/_layouts/WopiFrame.aspx?sourcedoc={{row.entity.matterClientUrl}}/{{row.entity.matterGuid}}_OneNote/{{row.entity.matterName}}/{{row.entity.matterGuid}}.onetoc2' target='_blank'>
-                        {{grid.appScope.vm.matterConfigContent.EllipsisLink3Text}}
-                    </a>
-                </li>
-                <li class='ms-ContextualMenu-item' ng-click='row.entity.MatterInfo===undefined?grid.appScope.vm.PinMatter(row):grid.appScope.vm.UnpinMatter(row)'><a title={{row.entity.MatterInfo===undefined?grid.appScope.vm.matterConfigContent.ECBMenu4Title:grid.appScope.vm.matterConfigContent.ECBMenu5Title}} class='ECBItem ms-ContextualMenu-link upload'>{{row.entity.MatterInfo===undefined?grid.appScope.vm.matterConfigContent.EllipsisLink4Text:grid.appScope.vm.matterConfigContent.EllipsisLink5Text}}</a></li>
+            <a class='btn-link dropdown-toggle ms-Icon ms-Icon--ellipsis ellipsis UiGrid-a' type='button' ng-click="grid.appScope.vm.checkUrlExists(row.entity)" dropdown data-toggle='dropdown'></a>
```
Comment (premchandpl) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Web/wwwroot/app/matter/MatterTemplates/MatterCellTemplate.html]: to add functionality for one note url existence

File: tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Web/wwwroot/app/matter/matters.controller.js [hunk 14]
```diff
@@ -351,6 +354,36 @@
                 });
             }
 
+            //Callback function for Onenote Url Exists
```
Comment (premchandpl) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Web/wwwroot/app/matter/matters.controller.js]: functionality for one note url existence

File: tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/SharePointInfo/ExternalSharing.cs [hunk 15]
```diff
@@ -110,35 +109,60 @@ private GenericResponseVM SendExternalNotification(MatterInformationVM matterInf
...
-                var users = new List<UserRoleAssignment>();
-                UserRoleAssignment userRole = new UserRoleAssignment();
-                userRole.UserId = externalEmail;
+            {
+                string roleId = "";
                 switch (permission.ToLower())
                 {
                     case "full control":
-                        userRole.Role = SharePoint.Client.Sharing.Role.Owner;
+                        roleId = "1073741829";
```
Comment (WilsonReddy) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/SharePointInfo/ExternalSharing.cs]: Define RoleValue enumeration and use it for assigning ids.

Comment (WilsonReddy) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/SharePointInfo/ExternalSharing.cs]: Define RoleValue enumeration and use it for assigning ids.

Comment (WilsonReddy) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/SharePointInfo/ExternalSharing.cs]: Define RoleValue enumeration and use it for assigning roles.

Comment (WilsonReddy) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/SharePointInfo/ExternalSharing.cs]: Define RoleValue enumeration and use it for assigning roles.

**Comment by premchandpl** [comment]
Will change that Wilson. Forgot to do that

Sent from my iPhone

On Nov 30, 2016, at 9:59 AM, Wilson Reddy Gajarla <notifications@github.com<mailto:notifications@github.com>> wrote:


@WilsonReddy commented on this pull request.

________________________________

In tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/SharePointInfo/ExternalSharing.cs<https://github.com/Microsoft/mattercenter/pull/590#pullrequestreview-10811416>:

>                  switch (permission.ToLower())
                 {
                     case "full control":
-                        userRole.Role = SharePoint.Client.Sharing.Role.Owner;
+                        roleId = "1073741829";


Define RoleValue enumeration and use it for assigning ids.

________________________________

In tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/SharePointInfo/ExternalSharing.cs<https://github.com/Microsoft/mattercenter/pull/590#pullrequestreview-10811416>:

>                  switch (permission.ToLower())
                 {
                     case "full control":
-                        userRole.Role = SharePoint.Client.Sharing.Role.Owner;
+                        roleId = "1073741829";


Define RoleValue enumeration and use it for assigning ids.

-

File: tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/SharePointInfo/ExternalSharing.cs [hunk 16]
```diff
@@ -110,35 +109,60 @@ private GenericResponseVM SendExternalNotification(MatterInformationVM matterInf
...
-                UserRoleAssignment userRole = new UserRoleAssignment();
-                userRole.UserId = externalEmail;
+            {
+                string roleId = "";                
                 switch (permission.ToLower())
                 {
-                    case "full control":
-                        userRole.Role = SharePoint.Client.Sharing.Role.Owner;
+                    case "full control":                        
+                        roleId = ((int)SPORoleIdMapping.FullControl).ToString();
```
Comment (premchandpl) [inline tree/master/cloud/src/solution/Microsoft.Legal.MatterCenter.Service/Microsoft.Legal.MatterCenter.Repository/SharePointInfo/ExternalSharing.cs]: Added enum for role id

**Pull request merged**